### PR TITLE
Issue/3470 unserializable items in the changes set make the crud handler get stuck in updating

### PR DESCRIPTION
**Pull request by Hugo-Inmanta**
# Description

Cast sets to string during JSON serialization

Part of #3470 

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [x] Attached issue to pull request
- [x] Changelog entry
- [x] Type annotations are present
- [x] Code is clear and sufficiently documented
- [x] No (preventable) type errors (check using make mypy or make mypy-diff)
- [x] Sufficient test cases (reproduces the bug/tests the requested feature)
- [x] Correct, in line with design
~~- [ ] End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )~~

# Reviewer Checklist:

- [ ] Sufficient test cases (reproduces the bug/tests the requested feature)
- [ ] Code is clear and sufficiently documented
- [ ] Correct, in line with design

### CODE REVIEW

File: src/inmanta/util.py [hunk 1]
```diff
@@ -310,6 +310,9 @@ def _custom_json_encoder(o: object) -> Union[ReturnTypes, "JSONSerializable"]:
         # Logs can push exceptions through RPC. Return a string representation.
         return str(o)
 
+    if isinstance(o, set):
+        return str(o)
```
Comment (wouterdb) [inline src/inmanta/util.py:314]: This I think is way overbroad: it changes the serialization behavior of the entire server.

File: src/inmanta/agent/handler.py [hunk 2]
```diff
@@ -909,7 +909,15 @@ def execute(self, ctx: HandlerContext, resource: resources.Resource, dry_run: Op
                     changes["purged"] = dict(desired=desired.purged, current=True)
 
             for field, values in changes.items():
-                ctx.add_change(field, desired=values["desired"], current=values["current"])
+                for status in ["desired", "current"]:
```
Comment (arnaudsjs) [inline src/inmanta/agent/handler.py:912]: Wouldn't it be better to move this logic into the `AttributeStateChange` class? (I might be missing something here)

**Review by arnaudsjs — CHANGES_REQUESTED** [review]

File: src/inmanta/agent/handler.py (hunk 2 above)
Comment (wouterdb) [reply src/inmanta/agent/handler.py:912]: or even in the add_change method, which is another codepath that needs this check.

**Review by wouterdb — CHANGES_REQUESTED** [review]

File: src/inmanta/agent/handler.py (hunk 2 above)
Comment (arnaudsjs) [reply src/inmanta/agent/handler.py:912]: @wouterdb I don't think that just adding it to the `add_change()` method would be sufficient. The way I see it, is the following: Any code path that updates the `self._changes` field of the `HandlerContext` class should be protected against unserializable things (add_change(), add_changes(), update_changes(), etc). All these code paths use the `AttributeStateChange()` class, so adding guard there would ensure that all code paths are safe. Do you agree?

**Comment by arnaudsjs** [comment]
Shouldn't we also protect the code path to the `log_msg()` method? Because that's the code path mentioned in the stack trace of the ticket.

File: changelogs/unreleased/3470-unserializable-items-in-the-changes-set-make-the-CRUD-handler-get-stuck-in-updating.yml [hunk 3]
```diff
@@ -0,0 +1,7 @@
+description: Cast items in the 'changes' set that aren't JSON-serializable to str.
```
Comment (arnaudsjs) [inline changelogs/unreleased/3470-unserializable-items-in-the-changes-set-make-the-CRUD-handler-get-stuck-in-updating.yml:1]: Keep in mind that changelog are intended to be read by end-user. A good changelog entry should describe what has been fixed without going in depth about the implementation details.
```suggestion
description: Fix bug that fails the CRUDHandler when a changed attribute is of type set.
```

File: src/inmanta/agent/handler.py [hunk 4]
```diff
@@ -921,8 +921,8 @@ def execute(self, ctx: HandlerContext, resource: resources.Resource, dry_run: Op
                         self.delete_resource(ctx, desired)
 
                 elif not desired.purged and len(changes) > 0:
-                    ctx.debug("Calling update_resource", changes=changes)
-                    self.update_resource(ctx, dict(changes), desired)
+                    ctx.debug("Calling update_resource", changes=ctx.changes)
+                    self.update_resource(ctx, dict(ctx.changes), desired)
```
Comment (arnaudsjs) [inline src/inmanta/agent/handler.py:925]: Why was this changed?

**Review by arnaudsjs — CHANGES_REQUESTED** [review]

File: src/inmanta/agent/handler.py (hunk 4 above)
Comment (Hugo-Inmanta) [reply src/inmanta/agent/handler.py:925]: Don't we want to pass the sanitized changes to these methods ?

Comment (wouterdb) [reply src/inmanta/agent/handler.py:925]: mhh, this is a delicate issue:

1. we have two mechanisms to pass on changes (via the context and via the return value). 
2. if I look at all the surrounding code, we only act on those that come from the return value
3. the server however uses the once from the context to trigger reloads and other events 
4. we correctly log what we act on 
5. the corresponding server log may contain many other things.

While the current behavior is not optimal, changing the input to update_resource is a MAJOR change to the behavior of the agent.

File: src/inmanta/agent/handler.py [hunk 5]
```diff
@@ -909,7 +909,7 @@ def execute(self, ctx: HandlerContext, resource: resources.Resource, dry_run: Op
                     changes["purged"] = dict(desired=desired.purged, current=True)
 
             for field, values in changes.items():
-                ctx.add_change(field, desired=values["desired"], current=values["current"])
+                ctx.add_change(field, desired=changes[field]["desired"], current=changes[field]["current"])
```
Comment (wouterdb) [inline src/inmanta/agent/handler.py:912]: Why was this changed?

**Review by wouterdb — CHANGES_REQUESTED** [review]

File: src/inmanta/agent/handler.py (hunk 5 above)
Comment (Hugo-Inmanta) [reply src/inmanta/agent/handler.py:912]: This is an artifact from a previous iteration, reverting the change now.
